src/main/scala/shine/DPIA/Compilation/ModuleGenerator.scala

```diff
@@ -0,0 +1,52 @@
+package shine.DPIA.Compilation
+
+import shine.DPIA
+import shine.DPIA.Phrases._
+import shine.DPIA.Types.{AccType, CommType, ExpType, PhraseType, TypeCheck}
+import util.compiler.DSL.run
+
+trait ModuleGenerator[FunDef <: DPIA.Compilation.FunDef] {
+  type Module
+  type CodeGenerator <: DPIA.Compilation.CodeGenerator
+
+  def makeFunDef(name: String): Phrase[_ <: PhraseType] => FunDef
+
+  def generateFunction(name: String,
+                       gen: CodeGenerator): Phrase[_ <: PhraseType] => Module =
+    makeFunDef(name) andThen funDefToModule(gen)
+
+  def funDefToModule(gen: CodeGenerator): FunDef => Module = funDef => {
+    val outParam = createOutputParam(funDef.body.t)
+
+    funDef.body |> (
+      run(TypeCheck(_: Phrase[ExpType])) andThen
+      rewriteToImperative(gen, funDef, outParam) andThen
+      imperativePasses(gen, funDef, outParam) andThen
+      generateCode(gen, funDef, outParam) andThen
+      makeModule(gen, funDef, outParam) )
+  }
+
+  def createOutputParam(outT: ExpType): Identifier[AccType]
+
+  def rewriteToImperative(gen: CodeGenerator,
+                          funDef: FunDef,
+                          outParam: Identifier[AccType]
+                         ): Phrase[ExpType] => Phrase[CommType]
+
+  type PhraseAfterPasses
+  def imperativePasses(gen: CodeGenerator,
+                       funDef: FunDef,
+                       outParam: Identifier[AccType]
+                      ): Phrase[CommType] => PhraseAfterPasses
+
+  type GeneratedCode
+  def generateCode(gen: CodeGenerator,
+                   funDef: FunDef,
+                   outParam: Identifier[AccType]
+                  ): PhraseAfterPasses => GeneratedCode
+
+  def makeModule(gen: CodeGenerator,
+                 funDef: FunDef,
+                 outParam: Identifier[AccType]
+                ): GeneratedCode => Module
```

Comment:
is it useful to separate `imperativePasses`, `generateCode` and `makeModule` here? I wonder if the `ModuleGenerator` API should not merge them into something like `imperativeToModule`, avoiding these arbitrary stages with abstract result types. Specific module generators can split this process into stages as they desire?

Reply:
One advantage is that we can independently refer (and invoke) the passes that are performed, e.g., by the `C.ModuleGenerator`.
This way it is easy to ensure that the e.g. `OpenCL.KernelGenerator` is performing the same passes (plus more) as the C generator.

If we mix them all up in one big step this aspect becomes harder.

Reply:
I agree that `OpenCL.KernelGenerator` wants to refer to some `C.ModuleGenerator` passes, but this does not require any trait. I believe that the trait should only contain methods that are useful/meaninful in the generic context. It does not look like anybody would use `mg: ModuleGenerator; mg.imperativePasses()`. They would not even know what it returns concretely. We can keep these methods inside the concrete generators and remove them from the generic trait?

Reply:
I see your point and https://github.com/rise-lang/shine/pull/92/commits/77cf7c004a082fb1ec3de42ff21e82a0f254def0 provides a simplified version of the ModuleGenerator trait.